### PR TITLE
[SessionManager] Skip claims creation if supplier operator balance is too low

### PR DESCRIPTION
**Pull request by red-0ne**
## Summary

This PR introduces filters in the `SubmitClaims` step to ensure that the `SessionManager` only submits claims it can afford to prove.

Before each submission, it checks the `SupplierOperator` balance and filters the claims based on the available funds.

Claims are sorted by descending profitability to maximize rewards.

## Issue

Since submitting a proof incurs a cost, a `SupplierOperator` without sufficient funds would be unable to submit the required proof.

This could result in the supplier's stake being slashed if a proof submission is required.

The `SessionManager` must verify that it can afford a proof submission before creating the corresponding claim.

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [ ] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Documentation**: `make docusaurus_start`; only needed if you make doc changes
- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: pkg/relayer/session/claim.go [hunk 1]
```diff
@@ -235,3 +240,75 @@ func (rs *relayerSessionsManager) goCreateClaimRoots(
 	failSubmitProofsSessionsCh <- failedClaims
 	claimsFlushedCh <- flushedClaims
 }
+
+// payableProofsSessionTrees returns the session trees that the supplier operator
+// can afford to claim.
+func (rs *relayerSessionsManager) payableProofsSessionTrees(
```
Comment (Olshansk) [inline pkg/relayer/session/claim.go:246]:
```suggestion
// Note that all sessionTrees are associated with the same supplier operator address.
func (rs *relayerSessionsManager) payableProofsSessionTrees(
```

File: pkg/relayer/session/claim.go [hunk 2]
```diff
@@ -235,3 +240,75 @@ func (rs *relayerSessionsManager) goCreateClaimRoots(
 	failSubmitProofsSessionsCh <- failedClaims
 	claimsFlushedCh <- flushedClaims
 }
+
+// payableProofsSessionTrees returns the session trees that the supplier operator
+// can afford to claim.
```
Comment (Olshansk) [inline pkg/relayer/session/claim.go:245]:
```suggestion
// can afford to claim (i.e. pay the fee for submitting a proof).
```

File: pkg/relayer/session/claim.go [hunk 3]
```diff
@@ -235,3 +240,75 @@ func (rs *relayerSessionsManager) goCreateClaimRoots(
...
+
+	supplierOperatorBalanceCoin, err := rs.bankQueryClient.GetBalance(
+		ctx,
+		sessionTrees[0].GetSupplierOperatorAddress().String(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort the session trees by the sum of the claim root to ensure that the
```
Comment (Olshansk) [inline pkg/relayer/session/claim.go:268]: This type of detail & business logic should be part of the godoc.  #PUC

File: pkg/relayer/session/claim.go [hunk 4]
```diff
@@ -235,3 +240,75 @@ func (rs *relayerSessionsManager) goCreateClaimRoots(
...
+		if errB != nil {
+			logger.With(
+				"session_id", a.GetSessionHeader().GetSessionId(),
+				"claim_root", fmt.Sprintf("%x", rootA),
+			).Error().Err(errB).Msg("failed to calculate sum of claim root, assuming 0")
+			sumB = 0
+		}
+
+		// Sort in descending order.
+		return int(sumB - sumA)
```
Comment (Olshansk) [inline pkg/relayer/session/claim.go:292]: Does this function have a unit test to make sure it works as expected?

File: pkg/relayer/session/claim.go [hunk 5]
```diff
@@ -235,3 +240,75 @@ func (rs *relayerSessionsManager) goCreateClaimRoots(
...
+	claimableSessionTrees := []relayer.SessionTree{}
+	for _, sessionTree := range sessionTrees {
+		// If the supplier operator can afford to claim the session, add it to the
+		// claimableSessionTrees slice.
+		if supplierOperatorBalanceCoin.IsGTE(*proofSubmissionFeeCoin) {
+			claimableSessionTrees = append(claimableSessionTrees, sessionTree)
+			newSupplierOperatorBalanceCoin := supplierOperatorBalanceCoin.Sub(*proofSubmissionFeeCoin)
+			supplierOperatorBalanceCoin = &newSupplierOperatorBalanceCoin
+			continue
+		}
```
Comment (Olshansk) [inline pkg/relayer/session/claim.go:304]: Should we break on else?

File: pkg/relayer/session/claim.go [hunk 6]
```diff
@@ -235,3 +240,75 @@ func (rs *relayerSessionsManager) goCreateClaimRoots(
...
+
+		logger.With(
+			"session_id", sessionTree.GetSessionHeader().GetSessionId(),
+			"supplier_operator_balance", supplierOperatorBalanceCoin,
+			"proof_submission_fee", proofSubmissionFeeCoin,
+		).Warn().Msg("supplier operator cannot afford to submit proof for claim, skipping")
+	}
+
+	return claimableSessionTrees, nil
```
Comment (Olshansk) [inline pkg/relayer/session/claim.go:313]: Thoughts on adding a warning log line like: "Can only afford X / N claims"?

File: pkg/relayer/session/session_test.go [hunk 7]
```diff
@@ -135,24 +337,28 @@ func requireProofCountEqualsExpectedValueFromProofParams(t *testing.T, proofPara
...
-	// Publish a mined relay to the minedRelaysPublishCh to insert into the session tree.
-	minedRelay := testrelayer.NewUnsignedMinedRelay(t, activeSession, supplierOperatorAddress)
-	minedRelaysPublishCh <- minedRelay
-
-	// The relayerSessionsManager should have created a session tree for the relay.
-	waitSimulateIO()
+	return blockPublishCh, minedRelaysPublishCh
+}
 
+// playClaimAndProofSubmissionBlocks simulates the block heights at which claims and proofs
```
Comment (Olshansk) [inline pkg/relayer/session/session_test.go:343]: Nice comment

File: pkg/relayer/session/session_test.go [hunk 8]
```diff
@@ -135,24 +337,28 @@ func requireProofCountEqualsExpectedValueFromProofParams(t *testing.T, proofPara
...
+// claims and proofs creation for the session number.
+func playClaimAndProofSubmissionBlocks(
+	t *testing.T,
+	sessionStartHeight, sessionEndHeight int64,
+	supplierOperatorAddress string,
+	blockPublishCh chan<- client.Block,
+) {
 	// Publish a block to the blockPublishCh to simulate non-actionable blocks.
-	sessionStartHeight := sessionHeader.GetSessionStartBlockHeight()
+	// NB: This does not need to be done for each service.
```
Comment (Olshansk) [inline pkg/relayer/session/session_test.go:353]:
```suggestion
	// NB: This only needs to be done once per block regardless of the number of services, claims or proofs.
```

File: pkg/relayer/session/session_test.go [hunk 9]
```diff
@@ -67,13 +72,209 @@ func requireProofCountEqualsExpectedValueFromProofParams(t *testing.T, proofPara
...
+}
+
+// waitSimulateIO sleeps for a bit to allow the relayer sessions manager to
+// process asynchronously. This effectively simulates I/O delays which would
+// normally be present.
+func waitSimulateIO() {
+	time.Sleep(50 * time.Millisecond)
+}
+
+func uPOKTCoin(amount int64) *sdktypes.Coin {
```
Comment (Olshansk) [inline pkg/relayer/session/session_test.go:263]: godoc please

File: pkg/relayer/session/session_test.go [hunk 10]
```diff
@@ -67,13 +72,209 @@ func requireProofCountEqualsExpectedValueFromProofParams(t *testing.T, proofPara
 			ServiceId:               service.Id,
 		},
 	}
-	sessionHeader := activeSession.GetHeader()
+	supplierOperatorAddress := sample.AccAddress()
+	// Set the supplier operator balance to be able to submit the expected number of proofs.
+	supplierOperatorBalance := prooftypes.DefaultParams().ProofSubmissionFee.Amount.Int64() * 2
+	supplierClientMap := testsupplier.NewClaimProofSupplierClientMap(ctx, t, supplierOperatorAddress, proofCount)
```
Comment (Olshansk) [inline pkg/relayer/session/session_test.go:78]:
```suggestion
	// Set the supplier operator balance to be able to submit the expected number of proofs.
	feePerProof := prooftypes.DefaultParams().ProofSubmissionFee.Amount.Int64()
	numExpectedProofs := 2
	supplierOperatorBalance := feePerProof * numExpectedProofs
	supplierClientMap := testsupplier.NewClaimProofSupplierClientMap(ctx, t, supplierOperatorAddress, proofCount)
```

File: pkg/relayer/session/session_test.go [hunk 11]
```diff
@@ -67,13 +72,209 @@ func requireProofCountEqualsExpectedValueFromProofParams(t *testing.T, proofPara
...
+	requireProofCountEqualsExpectedValueFromProofParams(t, proofParams, numExpectedProofs)
+}
+
+func TestRelayerSessionsManager_InsufficientBalanceForProofSubmission(t *testing.T) {
+	proofParams := prooftypes.DefaultParams()
+
+	// Set proof requirement threshold to a low enough value so a proof is always requested.
+	proofParams.ProofRequirementThreshold = uPOKTCoin(1)
+
+	// * Add 2 services with different CUPRs
```
Comment (Olshansk) [inline pkg/relayer/session/session_test.go:145]: Nice comment

File: pkg/relayer/session/session_test.go [hunk 12]
```diff
@@ -67,13 +72,209 @@ func requireProofCountEqualsExpectedValueFromProofParams(t *testing.T, proofPara
...
+
+	// The test is submitting a single claim. Having the proof request probability
+	// set to 0 and proof requirement threshold set to max uint64 results in no proofs
+	// being requested.
+	numExpectedProofs := 0
+
+	requireProofCountEqualsExpectedValueFromProofParams(t, proofParams, numExpectedProofs)
+}
+
+func TestRelayerSessionsManager_InsufficientBalanceForProofSubmission(t *testing.T) {
```
Comment (Olshansk) [inline pkg/relayer/session/session_test.go:139]: Great test, very easy to read!

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/relayer/session/claim.go (hunk 5 above)
Comment (red-0ne) [reply pkg/relayer/session/claim.go:304]: I wanted the log below to log all the missed claims.

File: pkg/relayer/session/claim.go (hunk 4 above)
Comment (red-0ne) [reply pkg/relayer/session/claim.go:296]: Yes [pkg/relayer/session/session_test.go#L220](https://github.com/pokt-network/poktroll/pull/817/files#diff-4118be5dadfdb12668216464d569818a98ab9694e4e9c9f505469317d9536516R220)

File: pkg/relayer/session/claim.go [hunk 13]
```diff
@@ -303,12 +307,18 @@ func (rs *relayerSessionsManager) payableProofsSessionTrees(
...
+		// Log a warning of any session that the supplier operator cannot afford to claim.
 		logger.With(
 			"session_id", sessionTree.GetSessionHeader().GetSessionId(),
 			"supplier_operator_balance", supplierOperatorBalanceCoin,
 			"proof_submission_fee", proofSubmissionFeeCoin,
 		).Warn().Msg("supplier operator cannot afford to submit proof for claim, skipping")
 	}
 
+	logger.Warn().Msgf(
+		"Supplier operator %q can only affort %d out of %d claims",
```
Comment (Olshansk) [inline pkg/relayer/session/claim.go:319]:
```suggestion
		"Supplier operator %q can only afford %d out of %d claims",
```

**Review by Olshansk — APPROVED** [review]

**Pull request merged**